### PR TITLE
Zak - OXO

### PR DESCRIPTION
**Pull request by zakarybk**
Needs AI.

### CODE REVIEW

**Comment by zakarybk** [comment]
I don't know how accurate the runtime from Travis is, as what I thought would be faster took 1 second longer.

File: .idea/comp110-worksheet-D.iml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (edpowley) [inline .idea/comp110-worksheet-D.iml:1]: It's good practice to add IDE-specific files to your `.gitignore` to exclude them from the repo.

File: oxo.py [hunk 2]
```diff
@@ -1,33 +1,99 @@
...
+        board -> Playable positions
+        board_width, board_height -> Board size
+        strike_combinations -> Each position for strike
+    """
+    board = {}
+
+    board_width = 3
+    board_height = 3
+
+    strike_combinations = [
```
Comment (edpowley) [inline oxo.py:17]: This approach is prone to copy-paste errors, and would be difficult to extend to larger boards. Could this list be initialised programmatically?

File: oxo.py [hunk 3]
```diff
@@ -1,33 +1,99 @@
...
+            self.current_moves += 1
+            return True
+        else:
+            return False
 
     def is_board_full(self):
         """ If there are still empty squares on the board, return False.
             If there are no empty squares, return True. """
-        raise NotImplementedError("TODO: implement is_board_full")
+        return self.current_moves >= self.total_moves
```
Comment (edpowley) [inline oxo.py:62]: Interesting approach but it should work! Though could just be `==` -- if `self.current_moves > self.total_moves` then presumably something has gone wrong

File: oxo.py [hunk 4]
```diff
@@ -1,33 +1,99 @@
...
 
     def get_winner(self):
         """ If a player has three in a row, return 1 or 2 depending on which player.
             Otherwise, return 0. """
-        raise NotImplementedError("TODO: implement get_winner")
+        for line in self.strike_combinations: # For each strike combination
+            strike_search = self.board[line[0][0], line[0][1]] # The first position for the strike
```
Comment (edpowley) [inline oxo.py:68]: Could have just used `self.board[line[0]]` here, since `line[0]` is a 2-tuple

File: oxo.py [hunk 5]
```diff
@@ -1,33 +1,99 @@
...
+            if strike_search == 0:
+                continue
+
+            # How many correct positions required in the current strike check to get a strike
+            parts_for_strike = len(line)
+            strike_complete = True
+
+            # For each position for strike line
+            for element in xrange(parts_for_strike):
+                x, y = line[element][0], line[element][1] # Get x and y from the current strike combination check
```
Comment (edpowley) [inline oxo.py:80]: This could have been written more succinctly (and more readably) as `for (x,y) in line:`

**Comment by edpowley** [comment]
Good -- correct and efficient implementation, however hard-coding of the winning lines is a maintainability issue. Try to initialise structures like these programmatically rather than by hard-coding.

**Pull request closed without merging**